pkg/shp/reactor/pod_watcher.go

```diff
@@ -0,0 +1,127 @@
+package reactor
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// PodWatcher a simple function orchestrator based on watching a given pod and reacting upon the
+// state modifications, should work as a helper to build business logic based on the build POD
+// changes.
+type PodWatcher struct {
+	ctx     context.Context
+	stopCh  chan bool       // stops the event loop execution
+	watcher watch.Interface // client watch instance
+
+	skipPODFn       SkipPODFn
+	onPoDAddedFn    OnPodEventFn
+	onPoDModifiedFn OnPodEventFn
+	onPoDDeletedFn  OnPodEventFn
+}
+
+// SkipPODFn a given pod instance is informed and expects a boolean as return. When true is returned
+// this container state processing is skipped completely.
+type SkipPODFn func(pod *corev1.Pod) bool
+
+// OnPodEventFn when a pod is modified this method handles the event.
+type OnPodEventFn func(pod *corev1.Pod) error
+
+// WithSkipPodFn sets the skip function instance.
+func (p *PodWatcher) WithSkipPodFn(fn SkipPODFn) *PodWatcher {
+	p.skipPODFn = fn
+	return p
+}
+
+// WithOnPodAddedFn sets the function executed when a pod is added.
+func (p *PodWatcher) WithOnPodAddedFn(fn OnPodEventFn) *PodWatcher {
+	p.onPoDAddedFn = fn
+	return p
+}
+
+// WithOnPodModifiedFn sets the funcion executed when a pod is modified.
+func (p *PodWatcher) WithOnPodModifiedFn(fn OnPodEventFn) *PodWatcher {
+	p.onPoDModifiedFn = fn
+	return p
+}
+
+// WithOnPodDeletedFn sets the funcion executed when a pod is modified.
+func (p *PodWatcher) WithOnPodDeletedFn(fn OnPodEventFn) *PodWatcher {
+	p.onPoDDeletedFn = fn
+	return p
+}
+
+// Start runs the event loop based on a watch instantiated against informed pod. In case of errors
+// the loop is interrupted.
+func (p *PodWatcher) Start() error {
```

Comment:
Pods are the common-denominator to what we need here. Both to follow logs and, in the near future, upload local data we need to know the pod, and also, if the status of the pod is susceptible. Bare in mind the [Build-Controller label the pods](https://github.com/shipwright-io/build/blob/5136c594b1adc5924bd063d0d6f4d5b264d6b99b/pkg/reconciler/buildrun/resources/taskrun.go#L224-L229) so it's trivial to find them without any additional objects.

In order to know the actual build status, we can check the `BuildRun` object status, which has just created to initialize the build process. And, we only need to inspect the `BuildRun` status when the pod-watch is done.

Reply:
I'm not worried about how hard it is to find the object.  I agree it is trivial.  

I'm worried about having to hit the api server again when we don't necessarily have to, since we are employing a watch here.  We have the object in hand.  Employing the use of watches/listers whenever possible is a performance convention that has been emphasized from the top on down in OCP.

And I'm not necessarily worried about the immediate scenarios at hand per se, but future ones.

Reply:
On this pull-request there's a single watch going on, a single watch over pods, so it's a single instance streaming events from the API-Server regardless.

Reply:
but there could be many users invoking `shp`

Reply:
I see what you mean, Gabe. Please consider the latest changes, on error the watcher will return the actual pod being handled during error.